sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/BufferingStreamObserver.java

```diff
@@ -41,25 +37,39 @@
  * becomes ready.
  */
 @ThreadSafe
-@SuppressWarnings({
-  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
-})
-public final class BufferingStreamObserver<T> implements StreamObserver<T> {
+public final class BufferingStreamObserver<T extends @NonNull Object> implements StreamObserver<T> {
```

Comment:
Just to make sure I'm understanding this class correctly, it is essentially working on two separate threads, right? One thread is the calls to the public API (i.e. the methods of `StreamObserver`, so `onNext`, `onCompleted`, and `onError`), while the other thread is the executor calling `drainQueue` which is the thread that actually does all the interaction with the underlying `outboundObserver`.

Am I understanding this class correctly?

Reply:
Yes.

The gRPC StreamObservers are typically not threadsafe across multiple threads so either you need to have threads synchronize or use a queue and have a dedicated thread writing.